src/economy/economy.cc

```diff
@@ -526,6 +526,76 @@ bool Economy::needs_ware_or_worker(DescriptionIndex const ware_or_worker_type) c
 	}
 }
 
+bool Economy::has_building(const DescriptionIndex di) const {
+	const std::string& name = owner().tribe().get_building_descr(di)->name();
+	for (const auto& bld : owner().get_building_statistics(di)) {
+		if (!bld.is_constructionsite) {
+			if (BaseImmovable* i = owner().egbase().map()[bld.pos].get_immovable()) {
+				if (i->descr().name() == name) {
+					if (dynamic_cast<const Building&>(*i).get_economy(type_) == this) {
+						return true;
+					}
+				}
+			}
+		}
+	}
+	return false;
+}
+
+ProductionSite* Economy::find_closest_occupied_productionsite(const Flag& base,
+                                                              const DescriptionIndex di,
+                                                              const bool check_inputqueues) {
+	const std::string& name = owner().tribe().get_building_descr(di)->name();
+	ProductionSite* best = nullptr;
+	uint32_t closest_dist = std::numeric_limits<uint32_t>::max();
+
+	auto check_building = [this, check_inputqueues](const ProductionSite& ps) {
```

Comment:
if you make this a Lamda function you can as well make it a parameter :-)

Can you make it a normal fucntion? or move it outside of this function?